src/linter.cc

```diff
@@ -24,93 +24,235 @@
+      if ( token.kind() == zetasql::ParseToken::KEYWORD ) {
+        if (!AllUpperCase(sql, token.GetLocationRange())) {
+          return absl::Status(
+            absl::StatusCode::kFailedPrecondition,
+            absl::StrCat("All keywords should be Uppercase, In character ",
+            std::to_string(token.GetLocationRange().start().GetByteOffset()),
+            " string should be: ", token.GetSQL()));
+        }
+      }
+    }
 
-    absl::Status parser_status = zetasql::ParseStatement(sql,
-            zetasql::ParserOptions(), &output);
+    return absl::OkStatus();
+}
 
-    if ( !parser_status.ok() )
-        return parser_status;
+absl::Status CheckCommentType(absl::string_view sql) {
+    bool includes_type1 = false;
+    bool includes_type2 = false;
+    bool inside_string = false;
+    
+    for (int i = 1; i<static_cast<int>(sql.size()); ++i) {
+        if (!inside_string && sql[i-1] == '-' && sql[i] == '-') {
+            includes_type1 = true;
+            // ignore the line
+            while ( i < static_cast<int>(sql.size()) &&
+                    sql[i] != '\n' ) {
+                ++i;
+            }
+        }
 
-    // KeywordExtractor(output->statement(), keywords).extract();
+        if (!inside_string && sql[i-1] == '/' && sql[i] == '/') {
+            includes_type2 = true;
+            // ignore the line
+            while ( i < static_cast<int>(sql.size()) &&
+                    sql[i] != '\n' ) {
+                ++i;
+            }
+        }
 
-    for (const zetasql::ASTNode *keyword : *keywords) {
-        if ( !allUpperCase(keyword) ) {
-            return absl::Status(
-                absl::StatusCode::kFailedPrecondition, "");
+        // ignore multiline comments
+        if (!inside_string && sql[i-1] == '/' && sql[i] == '*') {
+            // it will start checking after '/*' and after the iteration
+            // finished, the pointer 'i' will be just after '*/' (incrementation
+            // from the for statement is included)
+            i += 2;
+            while ( i < static_cast<int>(sql.size()) &&
+                    !(sql[i-1] == '*' && sql[i] == '/') ) {
+                ++i;
+            }
         }
+
+        if (sql[i] == '\'' || sql[i] == '"')
+            inside_string = !inside_string;
     }
 
+    if ( includes_type1 && includes_type2 )
+        return absl::Status(
+                absl::StatusCode::kFailedPrecondition,
+                absl::StrCat("either '//' or '--' should be used to ",
+                            "specify a comment"));
     return absl::OkStatus();
 }
 
+absl::Status ASTNodeRule::ApplyTo(absl::string_view sql) {
+    RuleVisitor visitor(rule_, sql);
+    absl::Status return_status = absl::OkStatus();
+    std::unique_ptr<zetasql::ParserOutput> output;
+
+    zetasql::ParseResumeLocation location =
+        zetasql::ParseResumeLocation::FromStringView(sql);
+    bool is_the_end = false;
+    while ( !is_the_end && return_status.ok() ) {
```

Comment:
with RETURN_IF_ERROR this while loop becomes shorter:

while (!is_the_end) {
  RETURN_IF_ERROR(zetasql::ParseNextScriptStatement(...));
  RETURN_IF_ERROR(output->statement()->TraverseNonRecursive(&visitor));
}

return visitor.GetResult();

Reply:
Fixed.